staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -449,15 +449,21 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 			if err != nil {
 				return nil, err
 			}
+			resourceVersion, err := w.versioner.ObjectResourceVersion(elem.Object)
+			if err != nil {
+				return nil, err
+			}
 			result[i] = &watchCacheEvent{
 				Type:            watch.Added,
 				Object:          elem.Object,
 				ObjLabels:       objLabels,
 				ObjFields:       objFields,
 				Key:             elem.Key,
-				ResourceVersion: w.resourceVersion,
+				ResourceVersion: resourceVersion,
 			}
 		}
+		// Sort by ResourceVersion so we send synthetic ADDED events in order of last update
+		sort.Slice(result, func(i, j int) bool { return result[i].ResourceVersion < result[j].ResourceVersion })
```

Comment:
This is more expensive than a properly typed one - since this is going to be a big sort would switch this to be a complied sort type.